<a href="https://colab.research.google.com/github/simplecod3r/introducao_analise_de_dados_e_estatistica/blob/main/lab02/Aula_Pr%C3%A1tica_PosDL_Lab02_ANOVA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Análise de Variância (ANOVA) em Python**

A análise de variância (ANOVA) é uma técnica estatística utilizada para determinar se há diferenças significativas entre as médias de dois ou mais grupos de dados. Em Python, podemos realizar ANOVA com a ajuda das bibliotecas statsmodels.

## **Iris Dataset**

Vamos considerar o conjunto de dados classico **Iris** muito utilizado para fins de aprendizado de máquina e análise de dados. Ele contém informações sobre três espécies diferentes de plantas Iris (Iris setosa, Iris virginica e Iris versicolor) e as medidas de quatro características de suas flores: comprimento e largura das pétalas e das sépalas.

O dataset contém 150 observações, sendo 50 para cada espécie de planta Iris. As medidas foram realizadas pelo botânico britânico Ronald Fisher em 1936, e o conjunto de dados é frequentemente utilizado para fins de classificação de espécies de flores Iris com base em suas medidas.

In [ ]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
import seaborn as sns

# Carregamos os dados com a lib seaborn
iris = sns.load_dataset('iris')

In [ ]:
iris.head(5)

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


In [ ]:
iris.describe()

,sepal_length,sepal_width,petal_length,petal_width
count,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333
std,0.828066,0.435866,1.765298,0.762238
min,4.300000,2.000000,1.000000,0.100000
25%,5.100000,2.800000,1.600000,0.300000
50%,5.800000,3.000000,4.350000,1.300000
75%,6.400000,3.300000,5.100000,1.800000
max,7.900000,4.400000,6.900000,2.500000


## Exemplo de código de ANOVA ONE-WAY

**Por que ANOVA one-way?**

A ANOVA one-way é usada quando:

* Há uma única variável independente (ou fator):

Neste caso, species é a única variável independente com três categorias.

* A variável dependente é contínua:

A variável sepal_length atende a este critério.

* O objetivo é comparar as médias entre os grupos:

A ANOVA testa se as médias de sepal_length para as três espécies são iguais ou não.

O objetivo deste estudo é determinar se a média do comprimento da sépala (sepal_length) difere entre as três espécies de flores do conjunto de dados Iris.

Para isso, utilizamos a Análise de Variância (ANOVA) One-Way, que permite comparar as médias de uma variável dependente contínua entre três ou mais grupos categóricos. Neste caso:

* Variável dependente: sepal_length (comprimento da sépala).
* Variável independente: species (espécie de flor: setosa, versicolor, virginica).

**Etapas do Processo**

1. Ajuste do Modelo ANOVA

O modelo ANOVA é ajustado utilizando a função ols() da biblioteca statsmodels.formula.api. Essa função realiza a regressão linear necessária para calcular os componentes da variância entre os grupos.

In [ ]:
model = ols('sepal_length ~ species', data=iris).fit()


2. Execução da ANOVA

Com o modelo ajustado, executamos a ANOVA usando a função anova_lm() da biblioteca statsmodels.stats.anova.

Aqui, utilizamos o argumento type=2, que indica o uso da soma de quadrados tipo II, recomendada quando os tamanhos dos grupos são diferentes.

In [ ]:
from statsmodels.stats.anova import anova_lm
anova_table = anova_lm(model, type=2)


3. **Interpretação dos Resultados**

A tabela resultante apresenta os seguintes componentes principais:

* df: Graus de liberdade para cada fonte de variação.
* sum_sq: Soma dos quadrados (variância explicada ou residual).
* mean_sq: Média dos quadrados (variância por grau de liberdade).
* F: Estatística F, que indica a razão entre a variância explicada e a variância residual.
*PR(>F): Valor-p, que indica se a diferença entre as médias é estatisticamente significativa

In [ ]:
print(anova_table)

             df     sum_sq    mean_sq           F        PR(>F)
species     2.0  63.212133  31.606067  119.264502  1.669669e-31
Residual  147.0  38.956200   0.265008         NaN           NaN


**Explicação da Tabela `anova_table`**

A tabela `anova_table` fornece informações sobre a decomposição da variância. Vamos entender cada coluna:

- **df (graus de liberdade)**: Representa a quantidade de informações independentes disponíveis para estimar os parâmetros.  
  - **species**: 2 graus de liberdade, pois há 3 categorias e utilizamos o cálculo:  
    
    Número de categorias - 1 = 3 - 1 = 2
  - **Residual**: 147 graus de liberdade restantes (\(150\) amostras - \(3\) categorias).

- **sum_sq (soma de quadrados)**: Representa a variância explicada (`species`) e a não explicada (`residuals`).

- **mean_sq (média dos quadrados)**: É a soma de quadrados dividida pelos graus de liberdade.

- **F (estatística F)**: Mede a proporção da variância explicada pela variável independente em relação à variância residual.  
  Um valor alto indica uma diferença significativa entre os grupos.

- **PR(>F)**: P-valor associado à estatística F.  
  Um valor muito pequeno (como \(1.67 \times 10^{-31}\)) indica evidências para rejeitar a hipótese nula, ou seja, as médias das espécies são significativamente diferentes.


## Post Hoc

Quando a ANOVA One-Way detecta diferenças significativas entre os grupos (com base no p-valor associado à estatística F), como é o caso aqui (p<0.05), é necessário realizar um teste post hoc. O teste post hoc ajuda a identificar especificamente quais grupos são diferentes entre si.

Neste caso, a análise ANOVA indica que existe uma diferença significativa entre as médias do comprimento da sépala (sepal_length) para as espécies, mas não informa entre quais espécies essas diferenças ocorrem. O post hoc resolve essa questão.

**Como fazer o post hoc?**

Um dos testes mais comuns para isso é o Teste de Tukey (Tukey HSD), que compara todas as combinações de pares de grupos, ajustando os valores-p para múltiplas comparações.

In [ ]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Aplicando o Teste de Tukey
posthoc = pairwise_tukeyhsd(endog=iris['sepal_length'], groups=iris['species'], alpha=0.05)

# Exibindo os resultados
print(posthoc)

   Multiple Comparison of Means - Tukey HSD, FWER=0.05   
  group1     group2   meandiff p-adj lower  upper  reject
---------------------------------------------------------
    setosa versicolor     0.93   0.0 0.6862 1.1738   True
    setosa  virginica    1.582   0.0 1.3382 1.8258   True
versicolor  virginica    0.652   0.0 0.4082 0.8958   True
---------------------------------------------------------


**Interpretação dos Resultados:**

O Teste de Tukey fornecerá uma tabela com:

* Os pares de grupos comparados.
* A diferença média entre os grupos.
* O intervalo de confiança.
* Um indicador se a diferença é significativa (
p<0.05).

Isso permitirá determinar quais pares de espécies (ex.: setosa vs. versicolor, versicolor vs. virginica, etc.) apresentam diferenças significativas no comprimento médio da sépala.

Interpretação:

* setosa vs versicolor: Diferença média de +0.93, com p<0.05. Isso significa que o comprimento médio da sépala em setosa é significativamente maior do que em versicolor.

* setosa vs virginica: Diferença média de +1.582, com p<0.05. O comprimento médio da sépala em setosa também é significativamente maior do que em virginica.

* versicolor vs virginica: Diferença média de +0.652, com p<0.05. O comprimento médio da sépala em versicolor é significativamente maior do que em virginica.

* Se meandiff > 0, significa que a média da primeira categoria (group1) é maior do que a média da segunda categoria (group2).

* Se meandiff < 0, significa que a média da primeira categoria (group1) é menor do que a média da segunda categoria (group2).

## Titanic Dataset

Vamos utilizar o conjunto de dados Titanic, que contém informações sobre os passageiros, como idade, sexo, classe, tarifa, entre outras variáveis. Para este exemplo de ANOVA Two-Way, analisaremos as seguintes variáveis:

**Variável dependente:**

* fare: Tarifa paga pelo passageiro (variável numérica contínua).

**Variáveis independentes (fatores):**

* class: Classe do passageiro (variável categórica com 3 níveis: primeira, segunda e terceira classe).
* sex: Sexo do passageiro (variável categórica com 2 níveis: masculino e feminino).

**O objetivo da análise é determinar:**

* Se a tarifa média varia entre as diferentes classes de passageiros.

* Se a tarifa média varia entre os sexos (masculino e feminino).

* Se há uma interação significativa entre as variáveis class e sex na determinação da tarifa média.

## Exemplo de código do ANOVA TWO-WAY

Para realizar um exemplo de ANOVA Two-Way, podemos utilizar um conjunto de dados que contenha duas variáveis categóricas e uma variável numérica contínua. Neste caso, vamos utilizar o conjunto de dados Titanic, que está disponível na biblioteca Seaborn do Python.

In [ ]:
titanic = sns.load_dataset('titanic')

In [ ]:
titanic.head(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [ ]:
# Mudamos o nome da coluna 'class' para 'classe' para evitar conflito com a palavra reservada 'class' do Python
titanic.rename(columns= {'class': 'classe'}, inplace=True)

# Criamos o modelo ANOVA Two-Way, onde a variável dependente é 'fare' (tarifa)

# E as variáveis independentes são 'classe' e 'sex', incluindo a interação entre elas

model = ols('fare ~ classe + sex + classe*sex', data=titanic).fit()

# Executamos a ANOVA com a decomposição dos quadrados tipo 2

anova_table = sm.stats.anova_lm(model, typ=2)

# Exibimos a tabela de resultados
print(anova_table)


                  sum_sq     df           F        PR(>F)
classe      7.347122e+05    2.0  242.620968  9.763645e-85
sex         3.174857e+04    1.0   20.968394  5.337586e-06
classe:sex  5.002816e+04    2.0   16.520591  9.034206e-08
Residual    1.339992e+06  885.0         NaN           NaN


**Explicação do Código:**

* Renomeação da Coluna:

Você renomeia a coluna class para classe, pois class é uma palavra reservada no Python e pode causar problemas.

* Modelo ANOVA:

O modelo ols('fare ~ classe + sex + classe*sex', data=titanic) ajusta um modelo de ANOVA Two-Way. O que está sendo feito aqui é:
fare: A variável dependente, que é a tarifa paga pelo passageiro.

classe e sex: São as variáveis independentes. Aqui estamos considerando os efeitos principais de cada variável.

classe*sex: A interação entre as variáveis classe e sex, o que é essencial em uma ANOVA Two-Way.

* ANOVA:

A função sm.stats.anova_lm(model, typ=2) realiza a análise de variância com a decomposição tipo 2, que é comum quando temos mais de um fator (no caso, classe e sex).

O tipo 2 é recomendado quando os fatores têm diferentes números de níveis e queremos avaliar os efeitos principais e as interações.
Resultados:

O código print(anova_table) vai exibir a tabela de ANOVA com os valores de soma de quadrados, graus de liberdade, estatísticas F e p-valores. Esses valores serão usados para interpretar os efeitos principais e as interações.

### Interpretando os resultados

**Interpretação Final:**

* Efeitos principais:

Tanto a variável classe quanto a variável sex têm efeitos significativos sobre a tarifa média paga pelos passageiros.

* Interação:
Há também uma interação significativa entre classe e sex. Isso sugere que a diferença nas tarifas entre as classes pode variar dependendo do sexo do passageiro (por exemplo, homens e mulheres podem ter diferenças nas tarifas dependendo da classe em que viajam).

## Post Hoc

Para complementar a análise, podemos realizar um teste post hoc para comparar as diferenças entre as classes e sexos. Neste caso, vamos utilizar o teste de Tukey. Podemos utilizar a função MultiComparison do statsmodels para criar um objeto que representa as comparações múltiplas entre os grupos, e em seguida aplicar o teste de Tukey com a função tukeyhsd.

In [ ]:
titanic['classe_sex'] = titanic['classe'].astype(str) + '_' + titanic['sex']

In [ ]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Criamos uma nova coluna juntando as variaveis independentes, isto é feito porque
# as funções não aceitam entradas bidimensionais só unidimensionais.
titanic['classe_sex'] = titanic['classe'].astype(str) + '_' + titanic['sex']


# Com esta função o endog recebe a variavel dependente, e o groups recebe a variavel independente.
tukey = pairwise_tukeyhsd(endog=titanic[['fare']],
                          groups=titanic[['classe_sex']], # groups = titanic[['classe', 'sex']]
                          alpha=0.05)

# Imprimindo os resultados
print(tukey)

         Multiple Comparison of Means - Tukey HSD, FWER=0.05         
    group1        group2    meandiff p-adj    lower    upper   reject
---------------------------------------------------------------------
 First_female    First_male -38.8997    0.0  -54.1513  -23.648   True
 First_female Second_female -84.1557    0.0 -101.2987 -67.0127   True
 First_female   Second_male  -86.384    0.0 -102.0599 -70.7081   True
 First_female  Third_female  -90.007    0.0 -104.7429 -75.2711   True
 First_female    Third_male -93.4642    0.0  -106.386 -80.5423   True
   First_male Second_female  -45.256    0.0  -61.4957 -29.0163   True
   First_male   Second_male -47.4843    0.0   -62.167 -32.8017   True
   First_male  Third_female -51.1073    0.0  -64.7819 -37.4328   True
   First_male    Third_male -54.5645    0.0  -66.2615 -42.8675   True
Second_female   Second_male  -2.2283 0.9989  -18.8672  14.4105  False
Second_female  Third_female  -5.8513 0.8968  -21.6077   9.9051  False
Second_female    Thi

Podemos ver que todas as combinações de grupos tiveram diferenças significativas nas médias de fare, exceto para a comparação entre Second_female e Second_male, e também para as comparações envolvendo Third_female e Third_male que não tiveram diferenças significativas.

No geral, o teste post hoc é útil quando a análise ANOVA indica que há diferenças significativas entre as médias, mas não nos diz quais grupos são diferentes uns dos outros. O teste post hoc nos ajuda a identificar exatamente quais grupos são diferentes e em que grau.





**Diferenças Significativas (onde reject = True):**

Há uma diferença significativa nas tarifas entre os seguintes pares de grupos:

* First_female vs. First_male: A diferença média é de -38.90, indicando que as mulheres na primeira classe pagam significativamente menos que os homens.

### Tabela Resumo: Diferenças entre One-Way e Two-Way ANOVA

| Característica           | One-Way ANOVA                     | Two-Way ANOVA                          |
|--------------------------|------------------------------------|----------------------------------------|
| **Número de fatores**    | 1 (ex.: `species`)                | 2 (ex.: `species` e `fertilizer`)      |
| **Número de grupos**     | Duas ou mais categorias           | Duas ou mais categorias para cada fator |
| **Interação analisada**  | Não                               | Sim (interação entre os dois fatores)  |
| **Exemplo de pergunta**  | A média de `sepal_length` varia entre espécies? | A média de `sepal_length` varia entre espécies, fertilizantes ou sua interação? |
